agents/components/MemberInvites.js

```diff
@@ -5,76 +5,95 @@ import { connect as connectFela } from 'react-fela'
 
-function renderMembers ({ fields, meta: { error, submitFailed }, formProps }) {
-  const { memberVals } = formProps
-  // TODO: currently this is an anti-pattern as it occurs within the render cycle
-  // TODO: mikey's idea was to not push state until the first edit to the empty row
-  if (memberVals) {
-    const memberKeys = Object.keys(memberVals)
-    if (memberKeys.length > 0) {
-      const lastMember = memberVals[memberKeys[memberKeys.length - 1]]
-      if (Object.keys(lastMember).length > 0) {
-        fields.push({})
-      }
-    }
-  }
-  return (
-    <div>
-      {submitFailed && error && <span>{error}</span>}
-      {fields.map((member, index) => (
-        <div key={index}>
-          <Field
-            name={`${member}.name`}
+class MemberInviteFields extends React.Component {
+  render () {
+    const { fields } = this.props
+    return (
+      <div ref='members'>
+        {fields.map((member, index) => (
+          <div key={index} className='member'>
+            <Field
+              name={`${member}.name`}
+              className='name'
+              floatingLabelText='Name'
+              component={TextField}
+            />
+            <Field
+              name={`${member}.email`}
+              className='email'
+              floatingLabelText='Email'
+              component={TextField}
+            />
+            <Field
+              name={`${member}.role`}
+              className='role'
+              floatingLabelText='Role'
+              component={TextField}
+            />
+            <Button type='button' onClick={() => fields.remove(index)}>Remove Member</Button>
+          </div>
+        ))}
+        <div key={fields.length}>
+          <MuiTextField
             floatingLabelText='Name'
-            component={TextField}
+            onChange={this.pushNext('name')}
           />
-          <Field
-            name={`${member}.email`}
+          <MuiTextField
             floatingLabelText='Email'
-            component={TextField}
+            onChange={this.pushNext('email')}
           />
-          <Field
-            name={`${member}.role`}
+          <MuiTextField
             floatingLabelText='Role'
-            component={TextField}
+            onChange={this.pushNext('role')}
           />
-          <Button type='button' onClick={() => fields.remove(index)}>Remove Member</Button>
         </div>
-      )
-    )}
-      <Button type='button' onClick={() => fields.push({})}>Add Member</Button>
+      </div>
+    )
+  }
+
+  pushNext (fieldName) {
+    const { fields } = this.props
+    return (ev, value) => {
+      fields.push({ [fieldName]: value })
+      const { members } = this.refs
+      // HACK (mw) how can we clean this up?
+      // keep focus after we push next member field
+      // because react elements will change underneath.
+      // also there's enough of a stutter that when
+      // typing you will lose some characters.
+      setTimeout(() => {
```

Comment:
in your approach, you pre-emptively add an empty field to the redux form state, then let it render as normal (even though it's empty). this means the react element tree works as normal, since the "empty row" is part of the `map(fields, renderField)` like any other row.

in my approach, i separate the "empty row" from the normal rows (for example, i don't have a remove button here because how can you remove something that doesn't exist yet :stuck_out_tongue_winking_eye:), so the react element tree is different, which causes react to change the underlying dom elements.

if we can't figure out this focus problem in a graceful way, i think we should continue with your approach but find a way to gracefully handle the "next empty row" being part of the state. (we need to handle this whenever we "save" the state, at least so the empty row state is isolated to this component.)